File: distribution/python/dqops/client/models/collect_statistics_on_table_queue_job_parameters.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.statistics_data_scope import StatisticsDataScope
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.collect_statistics_result import CollectStatisticsResult
    from ..models.physical_table_name import PhysicalTableName
    from ..models.statistics_collector_search_filters import (
        StatisticsCollectorSearchFilters,
    )


T = TypeVar("T", bound="CollectStatisticsOnTableQueueJobParameters")


@_attrs_define
class CollectStatisticsOnTableQueueJobParameters:
    """
    Attributes:
        connection (Union[Unset, str]): The name of the target connection.
        max_jobs_per_connection (Union[Unset, int]): The maximum number of concurrent 'run checks on table' jobs that
            can be run on this connection. Limits the number of concurrent jobs.
        table (Union[Unset, PhysicalTableName]):
        statistics_collector_search_filters (Union[Unset, StatisticsCollectorSearchFilters]):
        data_scope (Union[Unset, StatisticsDataScope]):
        samples_limit (Union[Unset, int]): The default limit of column samples that are collected.
        configure_table (Union[Unset, bool]): Turns on a special mode of collecting statistics that will configure the
            timestamp and ID columns. It should be used only during the first statistics collection.
        dummy_sensor_execution (Union[Unset, bool]): Boolean flag that enables a dummy statistics collection (sensors
            are executed, but the statistics results are not written to the parquet files).
        collect_statistics_result (Union[Unset, CollectStatisticsResult]): Returns the result with the summary of the
            statistics collected.
    """

    connection: Union[Unset, str] = UNSET
    max_jobs_per_connection: Union[Unset, int] = UNSET
    table: Union[Unset, "PhysicalTableName"] = UNSET
    statistics_collector_search_filters: Union[
        Unset, "StatisticsCollectorSearchFilters"
    ] = UNSET
    data_scope: Union[Unset, StatisticsDataScope] = UNSET
    samples_limit: Union[Unset, int] = UNSET
    configure_table: Union[Unset, bool] = UNSET
    dummy_sensor_execution: Union[Unset, bool] = UNSET
    collect_statistics_result: Union[Unset, "CollectStatisticsResult"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        connection = self.connection
        max_jobs_per_connection = self.max_jobs_per_connection
        table: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.table, Unset):
            table = self.table.to_dict()

        statistics_collector_search_filters: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.statistics_collector_search_filters, Unset):
            statistics_collector_search_filters = (
                self.statistics_collector_search_filters.to_dict()
            )

        data_scope: Union[Unset, str] = UNSET
        if not isinstance(self.data_scope, Unset):
            data_scope = self.data_scope.value

        samples_limit = self.samples_limit
        configure_table = self.configure_table
        dummy_sensor_execution = self.dummy_sensor_execution
        collect_statistics_result: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.collect_statistics_result, Unset):
            collect_statistics_result = self.collect_statistics_result.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if connection is not UNSET:
            field_dict["connection"] = connection
        if max_jobs_per_connection is not UNSET:
            field_dict["max_jobs_per_connection"] = max_jobs_per_connection
        if table is not UNSET:
            field_dict["table"] = table
        if statistics_collector_search_filters is not UNSET:
            field_dict["statistics_collector_search_filters"] = (
                statistics_collector_search_filters
            )
        if data_scope is not UNSET:
            field_dict["data_scope"] = data_scope
        if samples_limit is not UNSET:
            field_dict["samples_limit"] = samples_limit
        if configure_table is not UNSET:
            field_dict["configure_table"] = configure_table
        if dummy_sensor_execution is not UNSET:
            field_dict["dummy_sensor_execution"] = dummy_sensor_execution
        if collect_statistics_result is not UNSET:
            field_dict["collect_statistics_result"] = collect_statistics_result

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.collect_statistics_result import CollectStatisticsResult
        from ..models.physical_table_name import PhysicalTableName
        from ..models.statistics_collector_search_filters import (
            StatisticsCollectorSearchFilters,
        )

        d = src_dict.copy()
        connection = d.pop("connection", UNSET)

        max_jobs_per_connection = d.pop("max_jobs_per_connection", UNSET)

        _table = d.pop("table", UNSET)
        table: Union[Unset, PhysicalTableName]
        if isinstance(_table, Unset):
            table = UNSET
        else:
            table = PhysicalTableName.from_dict(_table)

        _statistics_collector_search_filters = d.pop(
            "statistics_collector_search_filters", UNSET
        )
        statistics_collector_search_filters: Union[
            Unset, StatisticsCollectorSearchFilters
        ]
        if isinstance(_statistics_collector_search_filters, Unset):
            statistics_collector_search_filters = UNSET
        else:
            statistics_collector_search_filters = (
                StatisticsCollectorSearchFilters.from_dict(
                    _statistics_collector_search_filters
                )
            )

        _data_scope = d.pop("data_scope", UNSET)
        data_scope: Union[Unset, StatisticsDataScope]
        if isinstance(_data_scope, Unset):
            data_scope = UNSET
        else:
            data_scope = StatisticsDataScope(_data_scope)

        samples_limit = d.pop("samples_limit", UNSET)

        configure_table = d.pop("configure_table", UNSET)

        dummy_sensor_execution = d.pop("dummy_sensor_execution", UNSET)

        _collect_statistics_result = d.pop("collect_statistics_result", UNSET)
        collect_statistics_result: Union[Unset, CollectStatisticsResult]
        if isinstance(_collect_statistics_result, Unset):
            collect_statistics_result = UNSET
        else:
            collect_statistics_result = CollectStatisticsResult.from_dict(
                _collect_statistics_result
            )

        collect_statistics_on_table_queue_job_parameters = cls(
            connection=connection,
            max_jobs_per_connection=max_jobs_per_connection,
            table=table,
            statistics_collector_search_filters=statistics_collector_search_filters,
            data_scope=data_scope,
            samples_limit=samples_limit,
            configure_table=configure_table,
            dummy_sensor_execution=dummy_sensor_execution,
            collect_statistics_result=collect_statistics_result,
        )

        collect_statistics_on_table_queue_job_parameters.additional_properties = d
        return collect_statistics_on_table_queue_job_parameters

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/execution/statistics/jobs/CollectStatisticsOnTableQueueJobParameters.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.execution.statistics.jobs;

import com.dqops.data.statistics.factory.StatisticsDataScope;
import com.dqops.execution.statistics.progress.SilentStatisticsCollectorExecutionProgressListener;
import com.dqops.execution.statistics.progress.StatisticsCollectorExecutionProgressListener;
import com.dqops.metadata.search.StatisticsCollectorSearchFilters;
import com.dqops.metadata.sources.PhysicalTableName;
import com.dqops.utils.exceptions.DqoRuntimeException;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;

/**
 * Parameters object for the run the statistics collection job for a separate child job that collects statistics for each table separately, in order to improve concurrency.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@EqualsAndHashCode(callSuper = false)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public class CollectStatisticsOnTableQueueJobParameters implements Cloneable {
    /**
     * The name of the target connection.
     */
    @JsonPropertyDescription("The name of the target connection.")
    private String connection;

    /**
     * The maximum number of concurrent 'run checks on table' jobs that can be run on this connection. Limits the number of concurrent jobs.
     * Null value means that there are no limits applied.
     */
    @JsonPropertyDescription("The maximum number of concurrent 'run checks on table' jobs that can be run on this connection. Limits the number of concurrent jobs.")
    private Integer maxJobsPerConnection;

    /**
     * The full name of a target table.
     */
    @JsonPropertyDescription("The full physical name (schema.table) of the target table.")
    private PhysicalTableName table;

    /**
     * Statistics collectors search filters that identify the type of statistics collector to run.
     */
    @JsonPropertyDescription("Statistics collectors search filters that identify the type of statistics collector to run.")
    private StatisticsCollectorSearchFilters statisticsCollectorSearchFilters;

    /**
     * The target scope of collecting statistics. Statistics can be collected for the entire or for each data stream separately.
     */
    @JsonPropertyDescription("The target scope of collecting statistics. Statistics can be collected for the entire table or for each data grouping separately.")
    private StatisticsDataScope dataScope = StatisticsDataScope.table;

    /**
     * The default limit of column samples that are collected.
     */
    @JsonPropertyDescription("The default limit of column samples that are collected.")
    private Integer samplesLimit;

    /**
     * Turns on a special mode of collecting statistics that will configure the timestamp and ID columns. It should be used only during the first statistics collection.
     */
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    @JsonPropertyDescription("Turns on a special mode of collecting statistics that will configure the timestamp and ID columns. It should be used only during the first statistics collection.")
    private boolean configureTable;

    /**
     * Progress listener that will receive events during the statistics collection.
     */
    @JsonIgnore
    @EqualsAndHashCode.Exclude
    private StatisticsCollectorExecutionProgressListener progressListener = new SilentStatisticsCollectorExecutionProgressListener();

    /**
     * Boolean flag that enables a dummy statistics collection (sensors are executed, but the statistics results are not written to the parquet files).
     */
    @JsonPropertyDescription("Boolean flag that enables a dummy statistics collection (sensors are executed, but the statistics results are not written to the parquet files).")
    private boolean dummySensorExecution;

    /**
     * The summary of the statistics collection job after if finished. Returns the number of collectors analyzed, columns analyzed, statistics results captured.
     */
    @JsonPropertyDescription("The summary of the statistics collection job after if finished. Returns the number of collectors analyzed, columns analyzed, statistics results captured.")
    private CollectStatisticsResult collectStatisticsResult;

    public CollectStatisticsOnTableQueueJobParameters() {
    }

    /**
     * Creates statistics collection run parameters.
     * @param connection The name of the target connection.
     * @param maxJobsPerConnection The maximum number of concurrent 'run checks on table' jobs that can be run on this connection. Limits the number of concurrent jobs.
     * @param table The full physical name (schema.table) of the target table.
     * @param statisticsCollectorSearchFilters Statistics collector search filters.
     * @param progressListener Progress listener to receive events during the statistics collector execution.
     * @param dataScope Statistics data scope to analyze - the whole table or each data stream separately.
     * @param samplesLimit The limit of the number of samples to collect.
     * @param configureTable The columns on the table should be configured.
     * @param dummySensorExecution True when it is a dummy run, only for showing rendered sensor queries.
     */
    public CollectStatisticsOnTableQueueJobParameters(String connection,
                                                      Integer maxJobsPerConnection,
                                                      PhysicalTableName table,
                                                      StatisticsCollectorSearchFilters statisticsCollectorSearchFilters,
                                                      StatisticsCollectorExecutionProgressListener progressListener,
                                                      StatisticsDataScope dataScope,
                                                      Integer samplesLimit,
                                                      boolean configureTable,
                                                      boolean dummySensorExecution) {
        this.connection = connection;
        this.maxJobsPerConnection = maxJobsPerConnection;
        this.table = table;
        this.statisticsCollectorSearchFilters = statisticsCollectorSearchFilters;
        this.progressListener = progressListener;
        this.dataScope = dataScope;
        this.samplesLimit = samplesLimit;
        this.configureTable = configureTable;
        this.dummySensorExecution = dummySensorExecution;
    }

    /**
     * Returns the name of the target connection.
     * @return Target connection name.
     */
    public String getConnection() {
        return connection;
    }

    /**
     * Sets the name of the target connection.
     * @param connection Target connection name.
     */
    public void setConnection(String connection) {
        this.connection = connection;
    }

    /**
     * Returns the limit of concurrent jobs (collect statistics on table) that can be executed in parallel on this connection.
     * @return Maximum number of concurrent statistics collection executions on this connection.
     */
    public Integer getMaxJobsPerConnection() {
        return maxJobsPerConnection;
    }

    /**
     * Sets the limit of concurrent jobs (collect statistics on table) that can be executed in parallel on this connection.
     * @param maxJobsPerConnection Maximum number of concurrent statistics collection executions on this connection.
     */
    public void setMaxJobsPerConnection(Integer maxJobsPerConnection) {
        this.maxJobsPerConnection = maxJobsPerConnection;
    }

    /**
     * Gets the full physical name (schema.table) of the target table.
     * @return Physical name of the target table.
     */
    public PhysicalTableName getTable() {
        return table;
    }

    /**
     * Sets the full physical name (schema.table) of the target table.
     * @param table Physical name of the target table.
     */
    public void setTable(PhysicalTableName table) {
        this.table = table;
    }

    /**
     * Returns the statistics collectors search filters.
     * @return Statistics collectors search filters.
     */
    public StatisticsCollectorSearchFilters getStatisticsCollectorSearchFilters() {
        return statisticsCollectorSearchFilters;
    }

    /**
     * Sets the statistics collectors search filters.
     * @param statisticsCollectorSearchFilters Statistics collectors search filters.
     */
    public void setStatisticsCollectorSearchFilters(StatisticsCollectorSearchFilters statisticsCollectorSearchFilters) {
        this.statisticsCollectorSearchFilters = statisticsCollectorSearchFilters;
    }

    /**
     * Progress listener that should be used to run the statistics collectors.
     * @return Statistics collectors progress listener.
     */
    public StatisticsCollectorExecutionProgressListener getProgressListener() {
        return progressListener;
    }

    /**
     * Sets the progress listener that should be used to run the statistics collectors.
     * @param progressListener Statistics collectors progress listener.
     */
    public void setProgressListener(StatisticsCollectorExecutionProgressListener progressListener) {
        this.progressListener = progressListener;
    }

    /**
     * Returns the statistics data scope (whole table or each data stream separately).
     * @return Data scope configuration.
     */
    public StatisticsDataScope getDataScope() {
        return dataScope;
    }

    /**
     * Sets the statistics data scope (whole table or each data stream separately).
     * @param dataScope Data scope configuration.
     */
    public void setDataScope(StatisticsDataScope dataScope) {
        this.dataScope = dataScope;
    }

    /**
     * Returns the limit of the column samples to capture.
     * @return The limit of columns samples.
     */
    public Integer getSamplesLimit() {
        return samplesLimit;
    }

    /**
     * Sets the limit of column samples to capture.
     * @param samplesLimit Column samples limit.
     */
    public void setSamplesLimit(Integer samplesLimit) {
        this.samplesLimit = samplesLimit;
    }

    /**
     * Returns true when the tables should be configured.
     * @return True when the tables should be configured.
     */
    public boolean isConfigureTable() {
        return configureTable;
    }

    /**
     * Sets the flag to configure the table.
     * @param configureTable True when the table should be configured.
     */
    public void setConfigureTable(boolean configureTable) {
        this.configureTable = configureTable;
    }

    /**
     * Returns true if it should be just a dummy run without actually running any queries.
     * @return true when it is a dummy run.
     */
    public boolean isDummySensorExecution() {
        return dummySensorExecution;
    }

    /**
     * Sets the flag if it should be just a dummy run without actually running any queries.
     * @param dummySensorExecution true when it is a dummy run.
     */
    public void setDummySensorExecution(boolean dummySensorExecution) {
        this.dummySensorExecution = dummySensorExecution;
    }

    /**
     * Returns the results of a finished job with the count of statistics that were collected.
     * @return The results of a finished job with the count of statistics that were collected.
     */
    public CollectStatisticsResult getCollectStatisticsResult() {
        return collectStatisticsResult;
    }

    /**
     * Sets the summary result of the statistics collection job, this method is called by the job when it finished successfully.
     * @param collectStatisticsResult The statistics collection result.
     */
    public void setCollectStatisticsResult(CollectStatisticsResult collectStatisticsResult) {
        this.collectStatisticsResult = collectStatisticsResult;
    }

    /**
     * Creates and returns a copy of this object.
     */
    @Override
    public CollectStatisticsOnTableQueueJobParameters clone() {
        try {
            return (CollectStatisticsOnTableQueueJobParameters)super.clone();
        }
        catch (CloneNotSupportedException ex) {
            throw new DqoRuntimeException("Clone not supported", ex);
        }
    }
}
